src/ray/util/counter_map.h

```diff
@@ -71,6 +79,13 @@ class CounterMap {
   /// to zero, the entry for the key is erased from the counter. It is not allowed for the
   /// count to be decremented below zero.
   void Decrement(const K &key, int64_t val = 1) {
+    // If value is 0, it is no-op and only registers the callback.
```

Comment:
these are necessary because of the bug I found (that was there, but it wasn't just caught). Basically when you have multi threads, there could be more than 1 get counter which breaks the invariant we have when writing this code.

To make it work, I explicitly supported "no-op mode" case.

Reply:
this is not thread safe (originally). The higher layer that uses this class is thread safe I believe